signer/src/main.rs

```diff
@@ -26,5 +85,25 @@ async fn main() {
 
     // run our app with hyper
     let listener = tokio::net::TcpListener::bind("0.0.0.0:8801").await.unwrap();
-    axum::serve(listener, app).await.unwrap();
+
+    // Subscribe to the signal channel so that we can catch shutdown events.
+    let mut signal = ctx.signal_subscribe();
+
+    // Start the server in its own task.
+    let handle = tokio::spawn(async { axum::serve(listener, app).await });
```

Comment:
That definitely works if `axum` is the only thing that needs to react to signals, but iirc there can only be one signal handler per signal per process, which is the last one "installed". Then we wouldn't be able to listen on the application-level, which would deprive the application of the ability to signal to the other components to shutdown gracefully (i.e. for libp2p we may want to send a "disconnect" message that triggers other peers to preemptively remove us to avoid unnecessary chatter and/or possibility of being penalized).

I guess the handler could still send a shutdown signal to the rest of the components, but then we're making the `axum` server task responsible for application termination signalling, which feels wrong...

Reply:
Oh yeah, I meant either have the future in the `with_graceful_shutdown` be the sender or a receiver. But I think you're right, it feels more natural for it to be a receiver.